corehq/apps/users/views/__init__.py

```diff
@@ -742,32 +706,94 @@ def invite_web_user_form(self):
+            # If user exists and has already requested access, just add them to the project
+            # Otherwise, send an invitation
+            create_invitation = True
             data = self.invite_web_user_form.cleaned_data
-            # create invitation record
-            data["invited_by"] = request.couch_user.user_id
-            data["invited_on"] = datetime.utcnow()
-            data["domain"] = self.domain
-            invite = DomainInvitation(**data)
-            invite.save()
-            invite.send_activation_email()
-            messages.success(request, "Invitation sent to %s" % invite.email)
+            domain_request = DomainRequest.by_email(self.domain, data["email"])
+            if domain_request is not None:
+                domain_request.is_approved = True
+                domain_request.save()
+                user = CouchUser.get_by_username(domain_request.email)
+                if user is not None:
+                    domain_request.send_approval_email()
+                    create_invitation = False
+                    user.add_as_web_user(self.domain, role=data["role"],
+                                         location_id=data["supply_point"], program_id=data["program"])
+                messages.success(request, "%s added." % data["email"])
+            else:
+                messages.success(request, "Invitation sent to %s" % data["email"])
+
+            if create_invitation:
+                data["invited_by"] = request.couch_user.user_id
+                data["invited_on"] = datetime.utcnow()
+                data["domain"] = self.domain
+                invite = DomainInvitation(**data)
+                invite.save()
+                invite.send_activation_email()
             return HttpResponseRedirect(reverse(
-                get_web_user_list_view(self.request).urlname,
+                ListWebUsersView.urlname,
                 args=[self.domain]
             ))
         return self.get(request, *args, **kwargs)
 
 
+class DomainRequestView(BasePageView):
+    urlname = "domain_request"
+    page_title = ugettext_lazy("Request Access")
+    template_name = "users/domain_request.html"
+    request_form = None
+
+    @property
+    def page_url(self):
+        return reverse(self.urlname, args=[self.request.domain])
+
+    @property
+    def page_context(self):
+        domain = Domain.get_by_name(self.request.domain)
+        if self.request_form is None:
+            self.request_form = DomainRequestForm(initial={'domain': domain.name})
+        return {
+            'domain': domain.name,
+            'domain_name': domain.display_name(),
+            'request_form': self.request_form,
+        }
+
+    def post(self, request, *args, **kwargs):
+        self.request_form = DomainRequestForm(request.POST)
+        if self.request_form.is_valid():
+            data = self.request_form.cleaned_data
+            with CriticalSection(["domain_request_%s" % data['domain']]):
```

Comment:
ah nice ... having just spent a while debugging an issue from duplicate domain objects being created with the same name, seeing code like this makes me happy